From 861f6492dbbd79bdcdd6213fb5d5da0e0cd22ab5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Erin=20Kalouskov=C3=A1?= <erin.kalouskova@nic.cz>
Date: Tue, 9 Dec 2025 16:46:26 +0100
Subject: [PATCH] Generate link signature for packages V2
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This implements same idea as patch by Karel Kočí <cynerd@email.cz>
which at some point broke by failing to enumerate external dynamic
symbols.

This patch, instead of using GNU Binutils' readelf and then filtering
and transducing it's output using awk, sed and grep uses LLVM
implementation of readelf which can output JSON we then process using
jq.

-- Original Commit Message --
This should solve most of the cases when new build of package with same
version is linked against newer or different version of library or that
new feature is linked in. This generates hash for all library and
dynamic symbols of all dynamic executables we are able to locate. Update
can check for this field and reinstall package if current install has
different signature. This comes from reality that single repository
snapshot has consistent link dependencies between packages and just
packages that are installed with same version from older snapshot can
break this dependency.
---
 include/package-pack.mk       |  7 +++++++
 scripts/gen-link-signature.sh | 30 ++++++++++++++++++++++++++++++
 2 files changed, 37 insertions(+)
 create mode 100755 scripts/gen-link-signature.sh

diff --git a/include/package-pack.mk b/include/package-pack.mk
index aa734f5af0..489c8d621b 100644
--- a/include/package-pack.mk
+++ b/include/package-pack.mk
@@ -85,6 +85,11 @@ ifneq ($(PKG_NAME),toolchain)
 		fi; \
 	)
   endef
+  define GenerateLinkSignature
+	LINK_SIGNATURE="$$$$( \
+		$(SCRIPT_DIR)/gen-link-signature.sh "$$(IDIR_$(1))"; \
+	)";
+  endef
 endif
 
 _addsep=$(word 1,$(1))$(foreach w,$(wordlist 2,$(words $(1)),$(1)),$(strip $(2) $(w)))
@@ -253,8 +258,10 @@ endif
 ifeq ($(CONFIG_USE_APK),)
 	mkdir -p $$(IDIR_$(1))/CONTROL
 	(cd $$(IDIR_$(1))/CONTROL; \
+		$(GenerateLinkSignature) \
 		( \
 			echo "$$$$CONTROL"; \
+			echo "LinkSignature: $$$$LINK_SIGNATURE"; \
 			printf "Description: "; echo "$$$$DESCRIPTION" | sed -e 's,^[[:space:]]*, ,g'; \
 		) > control; \
 		chmod 644 control; \
diff --git a/scripts/gen-link-signature.sh b/scripts/gen-link-signature.sh
new file mode 100755
index 0000000000..692fcd812a
--- /dev/null
+++ b/scripts/gen-link-signature.sh
@@ -0,0 +1,30 @@
+#!/usr/bin/env sh
+
+ELF_MAGIC="$(printf "\x7fELF")"
+
+[ -z "$*" ] && {
+	echo "$0: no directories or files specified" >&2
+	echo "usage: $0 [PATH...]" >&2
+	exit 1
+}
+
+find "$@" -type f | while IFS= read -r path; do
+	[ "$(head -c4 $path)" != "${ELF_MAGIC}" ] && continue
+	llvm-readelf \
+		--elf-output-style=JSON \
+		--dynamic-table \
+		--dyn-symbols \
+		"${path}"
+done \
+	| jq --slurp --compact-output '
+		reduce(.[][] | {
+			libraries : [.DynamicSection[] | select(.Type == "NEEDED") | .Library],
+			symbols   : [.DynamicSymbols[].Symbol
+			             | select(.Section.Name == "Undefined" and .Name.Name != "")
+			             | .Name.Name]
+		}) as $o
+			({libraries: {}, symbols: {}}; . * $o)
+			| map_values(.|sort)
+	' \
+	| md5sum \
+	| cut --delimiter=' ' --fields=1
-- 
2.52.0

